### PR TITLE
Encrypted Redis with Findex as a backend

### PR DESCRIPTION
**Pull request by bgrieder**
### Bug Fixes

- Documentation 

### Features

- Added `Redis-Findex` backend support so that the KMS can encrypt the KMS server data and indexes at the application level.

### CODE REVIEW

File: crate/server/src/database/redis/redis_with_findex.rs [hunk 1]
```diff
@@ -0,0 +1,601 @@
...
+    /// This method will fail if a `uid` is supplied
+    /// and an object with the same id already exists
+    async fn create_objects(
+        &self,
+        owner: &str,
+        objects: &[(Option<String>, Object, &HashSet<String>)],
+        _params: Option<&ExtraDatabaseParams>,
+    ) -> KResult<Vec<UniqueIdentifier>> {
+        // If the uid is not provided, generate a new one
+        let mut uids = vec![];
```
Comment (ThibsG) [inline crate/server/src/database/redis/redis_with_findex.rs:165]:
```suggestion
        let mut uids = Vec::with_capacity(objects.len());
```

File: crate/server/src/database/redis/redis_with_findex.rs [hunk 2]
```diff
@@ -0,0 +1,601 @@
...
+        let mut additions = HashMap::new();
+        let mut db_objects = HashMap::new();
+        for (uid, object, tags) in objects.iter() {
```
Comment (ThibsG) [inline crate/server/src/database/redis/redis_with_findex.rs:168]: `objects` is already an iterable slice
```suggestion
        for (uid, object, tags) in objects {
```

File: crate/server/src/database/redis/redis_with_findex.rs [hunk 3]
```diff
@@ -0,0 +1,601 @@
...
+fn intersect_all<I: IntoIterator<Item = HashSet<Location>>>(sets: I) -> HashSet<Location> {
+    let mut iter = sets.into_iter();
+    let first = iter.next().unwrap_or_default();
+    iter.fold(first, |acc, set| acc.intersection(&set).cloned().collect())
+}
+
+pub struct RedisWithFindex {
+    objects_db: Arc<ObjectsDB>,
+    permissions_db: PermissionsDB,
+    //TODO this Mutex should not be here; Findex needs to be changed to be thread-safe and not take &mut self
```
[the remark this answers is not in the document]
Comment (bgrieder) [reply crate/server/src/database/redis/redis_with_findex.rs:49]: No. I removed the outdated comment.

File: crate/server/src/database/redis/redis_with_findex.rs [hunk 4]
```diff
@@ -0,0 +1,601 @@
...
+        })
+    }
+}
+
+#[async_trait]
+impl Database for RedisWithFindex {
+    /// Return the filename of the database if supported
+    fn filename(&self, _group_id: u128) -> PathBuf {
+        PathBuf::from("")
+    }
```
[the remark this answers is not in the document]
Comment (bgrieder) [reply crate/server/src/database/redis/redis_with_findex.rs:93]: Yes agree but for some reason, the Database Trait was designed this way and MySQL and PostgreSQL actually return the same thing

Created an issue https://github.com/Cosmian/kms/issues/29

File: crate/server/src/database/redis/redis_with_findex.rs (hunk 1 above)
Comment (bgrieder) [reply crate/server/src/database/redis/redis_with_findex.rs:165]: Yes. Fixed

File: crate/server/src/database/redis/redis_with_findex.rs (hunk 2 above)
Comment (bgrieder) [reply crate/server/src/database/redis/redis_with_findex.rs:168]: Yes. Funny that Clippy misses that.
I guess, behind the scenes, calling `iter()` is exactly what happens

**Review by ThibsG — APPROVED** [review]

**Pull request merged**